File: java/MontecarloPI.java
import java.util.stream.IntStream;
import java.util.Random;

public class MontecarloPI {
    static IntStream repeat(int value, int times) {
        int[] tab = new int[times];
        for (int i = 0; i < times; i++) {
            tab[i] = value;
        }
        return IntStream.of(tab);
    }

    static int sample(int iters) {
        Random r = new Random(System.nanoTime());
        int inside = 0;
        for (int i = 0; i < iters; i++) {
            double x = r.nextDouble(),
                   y = r.nextDouble();

            if (x*x+y*y < 1.0) {
                inside++;
            }
        }
        return inside;
    }

    public static void main(String[] args) {
        final int precision   = Integer.parseInt(args[0]);
        final int concurrency = Integer.parseInt(args[1]);
        final int iters       = precision / concurrency;

        int result = repeat(iters, concurrency)
            .parallel()
            .map(MontecarloPI::sample)
            .reduce(0, Integer::sum);

        double pi = (double) result / precision * 4;
        System.out.println(pi);
    }
}

File: nim/montecarlopi.nim
import os, math, threadpool, strutils

proc sample(iters: int): int =
    for k in 1..iters:
        let x = random(1.0)
        let y = random(1.0)
        if x*x + y*y < 1.0:
            result += 1;


proc run(n: int, concurrency: int): float64 = 
    let iters = int(n / concurrency)
    var acc = newSeq[int](concurrency)
    parallel:
        for i in 0..acc.high:
            acc[i] = spawn sample(iters)

    for i in 0..acc.high:
        result += float(acc[i])

    result = 4.0 * result / float(n)

let args = commandLineParams()
echo(run(parseInt(args[0]), parseInt(args[1])))